### PR TITLE
Detect blanks with `exercise.blanks` opt, and add parse checking

### PR DESCRIPTION
**Pull request by rossellhayes**
This PR modifies `evaluate_exercise()` to check whether submitted code contains blanks and whether submitted code cannot be parsed.

If submitted code is unparsable, the R error message is rendered along with a message copied from `gradethis` explaining that the code could not be parsed and hinting at common causes of this problem.

One special case of unparsability is code that contains blanks. Often, tutorial authors will include blanks in their prompts or hints for students to fill in with valid code, e.g. `___(___, na.rm = TRUE)`. If submitted code contains blanks, evaluation proceeds as normal, but any feedback is replaced by a message explaining that the submitted code contains blanks and showing what blanks in the exercise look like. This applies even if the code containing blanks is parsable, e.g. if the blank is within a character string. To achieve this, the function `return_if_exercise_result()` is added within `evaluate_exercise()`. Each existing check (result check, code check, and error check) is wrapped within that function. This function initiates an early return if a result is signalled, but replaces the feedback with feedback about blanks if blanks are present in the user's code.

By default, any string of three or more underscores is considered to be a blank. This is the most common pattern in existing tutorials. But this definition can be changed with the `exercise.blanks` chunk option, which accepts a regex pattern to detect as blanks. For example, using `{r exercise-check, exercise.blanks = "\\.{3,}"}` as a chunk header would detect series of three or more dots as blanks.

Both unparsable and blanks messages are implemented with `i18next` keys, allowing them to be translated into the same language as the rest of the `learnr` interface. Instead of producing raw text when encountering unparsable code or code containing blanks, R generates a `span` tag with `data-i18n` attributes that are replaced with the correct localized text by the `i18next` JavaScript library on render. This PR will necessitate adding translations for `blank`, `blank_plural`, `exercisecontainsblank`, `pleasereplaceblank`, `unparsable`, `and`, and `or`.

### CODE REVIEW

File: tests/testthat/test-exercise.R [hunk 1]
```diff
@@ -626,3 +626,28 @@ test_that("env vars are protected from both user and author modification", {
...
+# unparsable input -----------------------------------------------------------
+
+test_that("evaluate_exercise() returns a message if code contains ___", {
+  ex     <- mock_exercise(user_code = '____("test")')
+  result <- evaluate_exercise(ex, new.env())
+  expect_match(result$error_message, "contains 1 blank")
+
+  ex     <- mock_exercise(user_code = '____("____")')
+  result <- evaluate_exercise(ex, new.env())
+  expect_match(result$error_message, "contains 2 blanks")
```
Comment (gadenbuie) [inline tests/testthat/test-exercise.R:638]: Maybe we should add a test here to make it clear that the code _has to be_ unparseable for the message about blanks, e.g. `user_code = 'cat("____")'` won't return the message about blanks.

File: R/exercise.R [hunk 2]
```diff
@@ -307,6 +307,14 @@ evaluate_exercise <- function(exercise, envir, evaluate_global_setup = FALSE) {
     eval(parse(text = exercise$global_setup), envir = envir)
   }
 
+  # Check if user code is parsable
+  parse_results <- tryCatch(
+    parse(text = exercise$code),
+    error = function(e) exercise_result_parse_error(e, exercise)
+  )
+
+  if (is_exercise_result(parse_results)) {return(parse_results)}
```
Comment (gadenbuie) [inline R/exercise.R:316]: If we decide to make this behavior default and always-on, we could even do this before evaluating the global setup code.

Comment (gadenbuie) [inline R/exercise.R:316]:
```suggestion
  if (is_exercise_result(parse_results)) {
    return(parse_results)
  }
```

**Review by gadenbuie — CHANGES_REQUESTED** [review]
This looks great, thank you!

One design question: In `gradethis` we give users an (undocumented) option as a way to turn this off or modify how unparseable code is handled. Should we also do that in learnr? Perhaps an option analogous to `exercise.error.check.code`, e.g. `exercise.error.parse.check`. (There may already be a better option in use in learnr.)

File: tests/testthat/test-exercise.R [hunk 3]
```diff
@@ -626,3 +626,28 @@ test_that("env vars are protected from both user and author modification", {
...
+})
+
+est_that("evaluate_exercise() returns a message if code is unparsable", {
```
Comment (gadenbuie) [inline tests/testthat/test-exercise.R:641]:
```suggestion
test_that("evaluate_exercise() returns a message if code is unparsable", {
```

**Comment by rossellhayes** [comment]
Are you thinking an `exercise.parse.error.check` would be implemented with global `options()` or with chunk options?

**Comment by gadenbuie** [comment]
1. Always run parse checker on submitted code before *any* evaluation
2. Follow the pattern of the `error-check` exercise checker
    - Use the `exercise.checker` function
    - Use the code from `*-parse-check` chunk as the `check_code` for that function
    - Or fall back to the global option `exercise.parse.check.code`
3. If parse failure, return the parse error message with additional friendly feedback

**Comment by rossellhayes** [comment]
Running into a few test failures on old releases because the new code uses `isFALSE()`, which wasn't available on R < 3.5.
I could:
1. use `backports`,
2. include `isFALSE()` as an internal function, or
3. refactor.

**Comment by rossellhayes** [comment]
Here's the current state of the checking:
1. Blank checking can be specified with `exercise.blank.check.code`. If it's `FALSE`, no blank checking is performed. If it's `NULL`, the default `exercise_blank_checker()` is called. Otherwise, it can be a function.
2. The default blank check is an `exercise.checker` that marks the exercise incorrect and tells the user how many blanks are still in their code. It detects both parsable (e.g. in strings) and unparsable blanks.
3. Blanks are specified with a regex in the `exercise.blanks` option. This defaults to `"_{3,}"`.
4. If there are no blanks, move to parse checking.
5. Parse checking can be specified with `exercise.blank.check.code`, using the same process as step 1.
6. The default parse check is an `exercise.checker` that marks the exercise as incorrect and an error if the user code cannot be parsed. It gives the same message as `gradethis` (you may have missed a blank, comma, end parenthesis, etc).
7. If the code parses, then the existing process begins: global setup is run, the exercise directory is created, etc.

File: R/i18n.R [hunk 4]
```diff
@@ -128,6 +128,18 @@ i18n_span <- function(key, ..., opts = NULL) {
   htmltools::HTML(format(x))
 }
 
+i18n_combine_words <- function(
+  words, and = c("and", "or"), before = "", after = before
+) {
+  and <- match.arg(and)
+  and <- paste0(" $t(text.", and, ") ")
```
Comment (gadenbuie) [inline R/i18n.R:135]: Very much a nit, but we can use old-school glue here.

```suggestion
  and <- sprintf(" $t(text.%s) ", and)
```

File: R/exercise.R [hunk 5]
```diff
@@ -327,34 +327,32 @@ evaluate_exercise <- function(
...
-    parse(text = exercise$code),
-    error = function(e) exercise_result_parse_error(e, exercise)
-  )
+  # Check if user code has unfilled blanks; if it does, early return
+  exercise_blanks <- exercise$options$exercise.blanks %||%
+    knitr::opts_chunk$get("exercise.blanks") %||%
+    "_{3,}"
 
-  if (is_exercise_result(parse_results)) {return(parse_results)}
+  if (isTruthy(exercise_blanks)) {
```
Comment (gadenbuie) [inline R/exercise.R:335]: Super nit, but let's prefix this so it's clear we didn't come up with our own version.

```suggestion
  if (shiny::isTruthy(exercise_blanks)) {
```

File: R/exercise.R [hunk 6]
```diff
@@ -415,9 +407,9 @@ evaluate_exercise <- function(
 }
 
 
-try_checker <- function(exercise, name, check_code, envir_result,
-                        evaluate_result, envir_prep, last_value,
-                        engine) {
+try_checker <- function(exercise, name, check_code = NULL, envir_result = NULL,
+                        evaluate_result = NULL, envir_prep, last_value = NULL,
+                        engine = exercise$engine) {
```
[the remark this answers is not in the document]
Comment (rossellhayes) [reply R/exercise.R:412]: Good catch, that's now a default argument. My original implementation of parse checking and blank checking called `try_checker()` with a different function, but with the three cases that still use it `exercise.checker` is there to access the different check codes within the exercise itself: `exercise$code_check`, `exercise$error_check` and `exercise$check`.

**Comment by rossellhayes** [comment]
These changes look good to me!

File: R/exercise.R [hunk 7]
```diff
@@ -336,34 +336,60 @@ evaluate_exercise <- function(
...
+  # the output may still be valid even if the user needs to fill in some blanks.
+  # Importantly, `blank_feedback` is `NULL` if no blanks are detected.
+  blank_feedback <- exercise_check_code_for_blanks(exercise)
+
+  here <- rlang::current_env()
+  return_if_exercise_result <- function(res) {
+    # early return if we've received an exercise result, but also replace the
+    # feedback with the blank feedback if any blanks were found
+    if (!is_exercise_result(res)) {
+      return(NULL)
```
Comment (gadenbuie) [inline R/exercise.R:361]:
```suggestion
      return()
```

File: R/exercise.R [hunk 8]
```diff
@@ -377,64 +403,64 @@ evaluate_exercise <- function(
...
-    return(rmd_results)
-  }
+  return_if_exercise_result(rmd_results)
 
+  # Check -------------------------------------------------------------------
   # Run the checker post-evaluation (for checking code results)
-  if (nzchar(exercise$check)) {
+  # Don't need to do exercise checking if there are blanks
+  checker_feedback <- NULL
+  if (is.null(blank_feedback) && nzchar(exercise$check)) {
```
Comment (gadenbuie) [inline R/exercise.R:430]: Return `blank_feedback` with output early instead of blending with post-eval check

**Review by gadenbuie — APPROVED** [review]

**Pull request merged**